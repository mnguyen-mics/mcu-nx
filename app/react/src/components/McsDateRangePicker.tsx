import * as React from 'react';
import { Button, DatePicker, Menu, Icon } from 'antd';
import moment from 'moment';
import { ClickParam } from 'antd/lib/menu';
import { Dropdown } from '../components/PopupContainers';
import withTranslations, { TranslationProps } from '../containers/Helpers/withTranslations';
import McsMoment, { convertMcsDateToMoment } from '../utils/McsMoment';

export interface McsDateRangeValue {
  from: McsMoment;
  to: McsMoment;
}

export interface McsDateRangePickerProps {
  values: McsDateRangeValue;
  onChange: (values: McsDateRangeValue) => void;
  format?: string;
}

interface McsDateRangePickerState {
  showRangePicker?: boolean;
}

interface Range {
  name: 'TODAY' | 'YESTERDAY' | 'LAST_7_DAYS' | 'LAST_30_DAYS';
  from: string;
  to: string;
}

const { RangePicker } = DatePicker;

const ranges: Range[] = [
  {
    name: 'TODAY',
    from : 'now',
    to: 'now',
  },
  {
    name: 'YESTERDAY',
    from : 'now-1d',
    to: 'now-1d',
  },
  {
    name: 'LAST_7_DAYS',
    from: 'now-7d',
    to: 'now',
  },
  {
    name: 'LAST_30_DAYS',
    from: 'now-30d',
    to: 'now',
  },
];

class McsDateRangePicker extends React.Component<McsDateRangePickerProps & TranslationProps, McsDateRangePickerState> {

  static defaultProps: Partial<McsDateRangePickerProps> = {
    format: 'YYYY-MM-DD',
  };

  state = {
    showRangePicker: false,
  };

  disableFutureDates(current: moment.Moment) {
    return current && current.valueOf() > Date.now();
  }

  getSelectedPresettedRange() {
    const { values, translations, format } = this.props;

    const selectedRange = ranges.find((range) => {
      return range.from === values.from.raw() && range.to ===  values.to.raw();
    });

    if (selectedRange) {
      return translations[selectedRange.name];
    }
    return `${convertMcsDateToMoment(values.from.raw())!.format(format)} ~ ${convertMcsDateToMoment(values.to.raw())!.format(format)}`;

  }

  handleDatePickerMenuChange = (dates: [moment.Moment, moment.Moment]) => {
    const { onChange } = this.props;

    this.setState({ showRangePicker: false });

    onChange({
      from: new McsMoment(dates[0].format('YYYY-MM-DD')),
      to: new McsMoment(dates[1].format('YYYY-MM-DD')),
    });
  }

  handleDropdownMenuClick = (param: ClickParam) => {
    const { onChange } = this.props;

    if (param.key === 'CUSTOM') {
      this.setState({
        showRangePicker: true,
      });
      return;
    }

    this.setState({ showRangePicker: false });

    const selectedRange = ranges.find(element => {
      return element.name.toLowerCase() === param.key.toLowerCase();
    });

    onChange({
      from: new McsMoment(selectedRange!.from),
      to:  new McsMoment(selectedRange!.to),
    });
  }

  onDatePickerOpenChange = () => {
    this.setState({
      showRangePicker: false,
    });
  }

  renderRangesDropdown() {
    const { translations } = this.props;

    const menu = (
      <Menu onClick={this.handleDropdownMenuClick}>
        <Menu.ItemGroup title={translations.LOOKBACK_WINDOW}>
          {
            ranges.map((item) => {
              return (this.getSelectedPresettedRange() === translations[item.name]
                ? null
                : <Menu.Item key={item.name}>{translations[item.name]}</Menu.Item>
              );
            })
          }
          <Menu.Item key="CUSTOM">{translations.CUSTOM}</Menu.Item>
        </Menu.ItemGroup>
      </Menu>
    );

    return (
      <Dropdown overlay={menu} trigger={['click']}>
        <Button>
          <Icon type="calendar" />
          {this.getSelectedPresettedRange()}
          <Icon type="down" />
        </Button>
      </Dropdown>
    );
  }

  render() {
    const { values } = this.props;
    const { showRangePicker } = this.state;
    const fromMoment = values.from.toMoment();
    const toMoment = values.to.toMoment();
    return showRangePicker === true
      ? (
          <RangePicker
            allowClear={false}
            onChange={this.handleDatePickerMenuChange}
            defaultValue={[fromMoment, toMoment]}
            disabledDate={this.disableFutureDates}
            onOpenChange={this.onDatePickerOpenChange}
            open={showRangePicker}
          />
        )
      : this.renderRangesDropdown();
  }
}

// TODO replace any with correct type
export default withTranslations(McsDateRangePicker) as any;
