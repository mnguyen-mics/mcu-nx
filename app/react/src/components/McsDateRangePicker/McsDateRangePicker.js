import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Dropdown, Button, DatePicker, Menu, Icon } from 'antd';
import moment from 'moment';

import { withTranslations } from '../../containers/Helpers';

const { RangePicker } = DatePicker;

const ranges = [{
  name: 'TODAY',
  dateRange: [moment(), moment()]
}, {
  name: 'YESTERDAY',
  dateRange: [moment().subtract(1, 'days'), moment()]
}, {
  name: 'LAST_7_DAYS',
  dateRange: [moment().subtract(7, 'days'), moment()]
}, {
  name: 'LAST_30_DAYS',
  dateRange: [moment().subtract(1, 'month'), moment()]
}];

const format = 'YYYY-MM-DD';

class McsDateRangePicker extends Component {

  constructor(props) {
    super(props);
    this.handleDropdownMenuClick = this.handleDropdownMenuClick.bind(this);
    this.handleDatePickerMenuChange = this.handleDatePickerMenuChange.bind(this);
    this.onDatePickerOpenChange = this.onDatePickerOpenChange.bind(this);
    this.state = {
      showRangePicker: false,
    };
  }

  handleDropdownMenuClick(e) {
    const {
      onChange
    } = this.props;

    if (e.key === 'CUSTOM') {
      this.setState({
        showRangePicker: true
      });
      return;
    }

    this.setState({
      showRangePicker: false,
    });

    const selectedRange = ranges.find(element => {
      return element.name.toLowerCase() === e.key.toLowerCase();
    });

    onChange({
      lookbackWindow: moment.duration(selectedRange.dateRange[1] - selectedRange.dateRange[0]),
      rangeType: 'relative',
      from: selectedRange.dateRange[0],
      to: selectedRange.dateRange[1]
    });
  }

  handleDatePickerMenuChange(dates) {
    const {
      onChange
    } = this.props;
    this.setState({
      showRangePicker: false,
    });
    onChange({
      rangeType: 'absolute',
      from: dates[0],
      to: dates[1],
      lookbackWindow: moment.duration(dates[1] - dates[0]),
    });
  }

  onDatePickerOpenChange() {
    this.setState({
      showRangePicker: false,
    });
  }

  getSelectedPresettedRange() {
    const {
      values,
      translations
    } = this.props;

    if (values.rangeType === 'absolute') {
      return `${values.from.format(format)} ~ ${values.to.format(format)}`;
    } else if (values.rangeType === 'relative') {

      const selectedRange = ranges.find((range) => {
        return Math.ceil(moment.duration(range.dateRange[1] - range.dateRange[0]).asSeconds()) === Math.ceil(values.lookbackWindow.asSeconds());
      });
      return selectedRange ? translations[selectedRange.name] : Math.ceil(values.lookbackWindow.asSeconds()).toString().concat(' ', translations.SECONDS);
    }
    return translations.ERROR;
  }


  renderRangesDropdown() {
    const {
      translations
    } = this.props;

    const menu = (
      <Menu onClick={(key) => this.handleDropdownMenuClick(key)}>
        <Menu.ItemGroup title={translations.LOOKBACK_WINDOW}>
          {
            ranges.map((item) => {
              return this.getSelectedPresettedRange() === translations[item.name] ? null : (<Menu.Item key={item.name}>{translations[item.name]}</Menu.Item>);
            })
          }
          <Menu.Item key="CUSTOM">{translations.CUSTOM}</Menu.Item>
        </Menu.ItemGroup>
      </Menu>
    );

    return (
      <Dropdown overlay={menu} trigger={['click']}>
        <Button>
          <Icon type="calendar" /> {this.getSelectedPresettedRange()} <Icon type="down" />
        </Button>
      </Dropdown>
    );
  }

  disableFutureDates(current) {
    return current && current.valueOf() > Date.now();
  }

  render() {
    const {
      values
    } = this.props;

    const {
      showRangePicker,
    } = this.state;

    return showRangePicker === true ? (<RangePicker
      allowClear={false}
      onChange={this.handleDatePickerMenuChange}
      defaultValue={[values.from, values.to]}
      disabledDate={current => this.disableFutureDates(current)}
      onOpenChange={this.onDatePickerOpenChange}
      open={showRangePicker}
    />) : this.renderRangesDropdown();
  }
}


McsDateRangePicker.propTypes = {
  values: PropTypes.shape({
    rangeType: PropTypes.string,
    lookbackWindow: PropTypes.object,
    from: PropTypes.object,
    to: PropTypes.object
  }).isRequired,
  onChange: PropTypes.func.isRequired,
  translations: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
};

McsDateRangePicker = withTranslations(McsDateRangePicker);

export default McsDateRangePicker;
