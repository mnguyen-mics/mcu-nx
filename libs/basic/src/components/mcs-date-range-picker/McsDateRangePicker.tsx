import React from 'react';
import { CalendarOutlined, ClockCircleOutlined, DownOutlined } from '@ant-design/icons';
import { Button, DatePicker, Menu } from 'antd';
import moment from 'moment';
import { MenuInfo } from 'rc-menu/lib/interface';
import { Dropdown } from '../popup-container/PopupContainer';
import McsMoment, { convertMcsDateToMoment } from '../../utils/McsMoment';
import OnOutsideClick from 'react-outclick';
import { AbstractMessages } from '../../utils/IntlHelper';

export interface McsDateRangeValue {
  from: McsMoment;
  to: McsMoment;
}

export interface McsDateRangePickerProps {
  className?: string;
  values: McsDateRangeValue;
  onChange: (values: McsDateRangeValue) => void;
  format?: string;
  disabled?: boolean;
  excludeToday?: boolean;
  startDate?: number;
  messages: McsDateRangePickerMessages;
}

interface McsDateRangePickerState {
  showDropdown?: boolean;
  datePickerOpen?: boolean;
  ranges: Range[];
  rangeType: 'ABSOLUTE' | 'RELATIVE';
}

export interface McsDateRangePickerMessages extends AbstractMessages {
  TODAY: string;
  YESTERDAY: string;
  LAST_7_DAYS: string;
  LAST_30_DAYS: string;
  LOOKBACK_WINDOW: string;
  CUSTOM: string;
  ABSOLUTE_TIME_RANGE: string;
  RELATIVE_TIME_RANGE: string;
}

interface Range {
  name: 'TODAY' | 'YESTERDAY' | 'LAST_7_DAYS' | 'LAST_30_DAYS';
  from: string;
  to: string;
}

const { RangePicker } = DatePicker;

type Props = McsDateRangePickerProps;
const defaultFormat = 'YYYY-MM-DD';

class McsDateRangePicker extends React.Component<Props, McsDateRangePickerState> {
  constructor(props: Props) {
    super(props);
    this.state = {
      showDropdown: false,
      ranges: [],
      rangeType: 'RELATIVE',
    };
  }

  componentDidMount() {
    const { excludeToday, startDate } = this.props;
    const ranges: Range[] = [
      {
        name: 'TODAY',
        from: 'now',
        to: 'now',
      },
      {
        name: 'YESTERDAY',
        from: 'now-1d',
        to: 'now-1d',
      },
    ];

    if (!startDate) {
      ranges.push(
        {
          name: 'LAST_7_DAYS',
          from: excludeToday ? 'now-8d' : 'now-7d',
          to: excludeToday ? 'now-1d' : 'now',
        },
        {
          name: 'LAST_30_DAYS',
          from: excludeToday ? 'now-31d' : 'now-30d',
          to: excludeToday ? 'now-1d' : 'now',
        },
      );
    }

    this.setState({
      ranges,
    });
  }

  disableDates = (current: moment.Moment) => {
    const { startDate } = this.props;
    return (
      current && (current.valueOf() > Date.now() || (!!startDate && current.valueOf() < startDate))
    );
  };

  getSelectedPresettedRange() {
    const { values, format, messages } = this.props;
    const { ranges } = this.state;

    const selectedRange = ranges.find(range => {
      return range.from === values.from.raw() && range.to === values.to.raw();
    });

    if (selectedRange) {
      return messages[selectedRange.name];
    }
    return `${convertMcsDateToMoment(values.from.raw())!.format(
      format || defaultFormat,
    )} ~ ${convertMcsDateToMoment(values.to.raw())!.format(format || defaultFormat)}`;
  }

  handleDatePickerMenuChange = (dates: [moment.Moment, moment.Moment]) => {
    const { onChange } = this.props;
    this.setState({
      rangeType: 'ABSOLUTE',
    });
    onChange({
      from: new McsMoment(dates[0].format('YYYY-MM-DD')),
      to: new McsMoment(dates[1].format('YYYY-MM-DD')),
    });
  };

  handleDropdownMenuClick = (param: MenuInfo) => {
    const { ranges } = this.state;
    const { onChange } = this.props;

    if (param.key !== 'CUSTOM') {
      this.setState({
        showDropdown: false,
        rangeType: 'RELATIVE',
      });

      const selectedRange = ranges.find(element => {
        return element.name.toLowerCase() === param.key.toString().toLowerCase();
      });

      onChange({
        from: new McsMoment(selectedRange!.from),
        to: new McsMoment(selectedRange!.to),
      });
    }
  };

  onDatePickerOpenChange = (open: boolean) => {
    this.setState({
      showDropdown: open,
      datePickerOpen: open,
    });
  };

  onDropdownClick = (event: React.MouseEvent<HTMLElement>) => {
    event.stopPropagation();
    this.setState({ showDropdown: true });
  };

  handleClick = () => {
    const { showDropdown, datePickerOpen } = this.state;

    if (showDropdown && !datePickerOpen) {
      this.setState({ showDropdown: false });
    }
  };

  renderRangesDropdown() {
    const { disabled, values, messages, className } = this.props;
    const { ranges, showDropdown, rangeType } = this.state;
    const fromMoment = values.from.toMoment();
    const toMoment = values.to.toMoment();
    const menu = (
      <Menu onClick={this.handleDropdownMenuClick} className={'mcs-dateRangePicker_menu'}>
        <Menu.ItemGroup
          title={'Absolute time range'}
          className={
            'mcs-dateRangePicker_menu_timeType mcs-dateRangePicker_menu_timeType--absolute'
          }
        >
          <Menu.Item
            className={'mcs-dateRangePicker_menu_item mcs-dateRangePicker_menu_item--rangePicker'}
            key='CUSTOM'
          >
            <RangePicker
              allowClear={false}
              className='mcs-dateRangePicker_rangePicker'
              onChange={this.handleDatePickerMenuChange}
              defaultValue={[fromMoment, toMoment]}
              disabledDate={this.disableDates}
              onOpenChange={this.onDatePickerOpenChange}
              value={[fromMoment, toMoment]}
            />
          </Menu.Item>
        </Menu.ItemGroup>
        <Menu.ItemGroup
          title={messages.RELATIVE_TIME_RANGE}
          className={'mcs-dateRangePicker_menu_timeType'}
        >
          {ranges.map(item => {
            return (
              <Menu.Item
                className={
                  this.getSelectedPresettedRange() === messages[item.name]
                    ? 'mcs-dateRangePicker_menu_item mcs-dateRangePicker_menu_item--selected'
                    : 'mcs-dateRangePicker_menu_item'
                }
                key={item.name}
              >
                {messages[item.name]}
              </Menu.Item>
            );
          })}
        </Menu.ItemGroup>
      </Menu>
    );

    return (
      <OnOutsideClick onOutsideClick={this.handleClick} display={'inline'}>
        <Dropdown
          className={`mcs-dateRangePicker ${className ? className : ''}`}
          overlay={menu}
          trigger={['click']}
          disabled={disabled}
          visible={showDropdown}
        >
          <Button onClick={this.onDropdownClick}>
            {rangeType === 'ABSOLUTE' ? <CalendarOutlined /> : <ClockCircleOutlined />}
            {this.getSelectedPresettedRange()}
            <DownOutlined />
          </Button>
        </Dropdown>
      </OnOutsideClick>
    );
  }

  render() {
    return this.renderRangesDropdown();
  }
}

export default McsDateRangePicker;
