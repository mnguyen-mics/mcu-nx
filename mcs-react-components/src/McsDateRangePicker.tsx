import * as React from 'react';
import { Dropdown, Button, DatePicker, Menu, Icon } from 'antd';
import moment from 'moment';
import { ClickParam } from 'antd/lib/menu';

import withTranslations, { TranslationProps } from './Helpers/withTranslations';

export interface McsDateRangeValue {
  rangeType: string;
  lookbackWindow?: moment.Duration;
  from?: moment.Moment;
  to?: moment.Moment;
}

export interface McsDateRangePickerProps {
  values: McsDateRangeValue;
  onChange: (values: McsDateRangeValue) => void;
  format?: string;
}

interface McsDateRangePickerState {
  showRangePicker?: boolean;
}

interface Range {
  name: 'TODAY' | 'YESTERDAY' | 'LAST_7_DAYS' | 'LAST_30_DAYS';
  from: moment.Moment;
  to: moment.Moment;
}

const { RangePicker } = DatePicker;

const currentTimeStampInt = parseInt(moment().format('X'), 10); // radix is needed
const numberOfSecondsInOneDay = 3600 * 24;

const ranges: Range[] = [
  {
    name: 'TODAY',
    from : moment.unix(currentTimeStampInt).startOf('day'),
    to: moment.unix(currentTimeStampInt + numberOfSecondsInOneDay).startOf('day'),
  },
  {
    name: 'YESTERDAY',
    from : moment.unix(currentTimeStampInt - numberOfSecondsInOneDay).startOf('day'),
    to: moment.unix(currentTimeStampInt).startOf('day'),
  },
  {
    name: 'LAST_7_DAYS',
    from: moment.unix(currentTimeStampInt - (7 * numberOfSecondsInOneDay)).startOf('day'),
    to: moment.unix(currentTimeStampInt + numberOfSecondsInOneDay).startOf('day'),
  },
  {
    name: 'LAST_30_DAYS',
    from: moment.unix(currentTimeStampInt - (30 * numberOfSecondsInOneDay)).startOf('day'),
    to: moment.unix(currentTimeStampInt + numberOfSecondsInOneDay).startOf('day'),
  },
];

class McsDateRangePicker extends React.Component<McsDateRangePickerProps & TranslationProps, McsDateRangePickerState> {

  static defaultProps: Partial<McsDateRangePickerProps> = {
    format: 'YYYY-MM-DD',
  };

  state = {
    showRangePicker: false,
  };

  disableFutureDates(current: moment.Moment) {
    return current && current.valueOf() > Date.now();
  }

  getSelectedPresettedRange() {
    const { values, translations, format } = this.props;

    if (values.rangeType === 'absolute') {
      return `${values.from!.format(format)} ~ ${values.to!.format(format)}`;
    } else if (values.rangeType === 'relative') {
      const selectedRange = ranges.find((range) => {
        const ceil1 = Math.ceil(moment
          .duration({ from: range.from, to: range.to })
          .asSeconds(),
        );
        return (ceil1 === Math.ceil(values.lookbackWindow!.asSeconds())
          && (range.to.unix() === (values.to ? values.to.unix() : null)));
      });

      return (selectedRange
        ? translations[selectedRange.name]
        : Math.ceil(values.lookbackWindow!.asSeconds()).toString().concat(' ', translations.SECONDS)
      );
    }

    return translations.ERROR;
  }

  handleDatePickerMenuChange = (dates: [moment.Moment, moment.Moment]) => {
    const { onChange } = this.props;

    this.setState({ showRangePicker: false });

    onChange({
      rangeType: 'absolute',
      from: dates[0],
      to: dates[1],
      lookbackWindow: moment.duration({from: dates[0], to: dates[1]}),
    });
  }

  handleDropdownMenuClick = (param: ClickParam) => {
    const { onChange } = this.props;

    if (param.key === 'CUSTOM') {
      this.setState({
        showRangePicker: true,
      });
      return;
    }

    this.setState({ showRangePicker: false });

    const selectedRange = ranges.find(element => {
      return element.name.toLowerCase() === param.key.toLowerCase();
    });

    onChange({
      lookbackWindow: moment.duration({ to: selectedRange!.to, from: selectedRange!.from }),
      rangeType: 'relative',
      from: selectedRange!.from,
      to: selectedRange!.to,
    });
  }

  onDatePickerOpenChange = () => {
    this.setState({
      showRangePicker: false,
    });
  }

  renderRangesDropdown() {
    const { translations } = this.props;

    const menu = (
      <Menu onClick={this.handleDropdownMenuClick}>
        <Menu.ItemGroup title={translations.LOOKBACK_WINDOW}>
          {
            ranges.map((item) => {
              return (this.getSelectedPresettedRange() === translations[item.name]
                ? null
                : <Menu.Item key={item.name}>{translations[item.name]}</Menu.Item>
              );
            })
          }
          <Menu.Item key="CUSTOM">{translations.CUSTOM}</Menu.Item>
        </Menu.ItemGroup>
      </Menu>
    );

    return (
      <Dropdown overlay={menu} trigger={['click']}>
        <Button>
          <Icon type="calendar" />
          {this.getSelectedPresettedRange()}
          <Icon type="down" />
        </Button>
      </Dropdown>
    );
  }

  render() {
    const { values } = this.props;
    const { showRangePicker } = this.state;

    return showRangePicker === true
      ? (
          <RangePicker
            allowClear={false}
            onChange={this.handleDatePickerMenuChange}
            defaultValue={[values.from!, values.to!]}
            disabledDate={this.disableFutureDates}
            onOpenChange={this.onDatePickerOpenChange}
            open={showRangePicker}
          />
        )
      : this.renderRangesDropdown();
  }
}

// TODO replace any with correct type
export default withTranslations(McsDateRangePicker) as any;
