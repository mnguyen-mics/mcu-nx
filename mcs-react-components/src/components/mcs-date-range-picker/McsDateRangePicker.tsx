import * as React from 'react';
import { CalendarOutlined, DownOutlined } from '@ant-design/icons';
import { Button, DatePicker, Menu } from 'antd';
import moment from 'moment';
import { MenuInfo } from '../../../node_modules/rc-menu/lib/interface';
import { Dropdown } from '../popupContainer/PopupContainer';
import McsMoment, { convertMcsDateToMoment } from '../../utils/McsMoment';
import { InjectedIntlProps, defineMessages, injectIntl } from 'react-intl';

export interface McsDateRangeValue {
  from: McsMoment;
  to: McsMoment;
}

export interface McsDateRangePickerProps {
  values: McsDateRangeValue;
  onChange: (values: McsDateRangeValue) => void;
  format?: string;
  disabled?: boolean;
  excludeToday?: boolean;
  startDate?: number;
}

interface McsDateRangePickerState {
  showRangePicker?: boolean;
  ranges: Range[];
}

interface Range {
  name: 'TODAY' | 'YESTERDAY' | 'LAST_7_DAYS' | 'LAST_30_DAYS';
  from: string;
  to: string;
}

const { RangePicker } = DatePicker;

const messages = defineMessages({
  TODAY: {
    id: 'components.mcsDateRangePicker.range.today',
    defaultMessage: 'Today',
  },
  YESTERDAY: {
    id: 'components.mcsDateRangePicker.range.yesterday',
    defaultMessage: 'Yesterday',
  },
  LAST_7_DAYS: {
    id: 'components.mcsDateRangePicker.range.last7days',
    defaultMessage: 'Last 7 days',
  },
  LAST_30_DAYS: {
    id: 'components.mcsDateRangePicker.range.last30days',
    defaultMessage: 'Last 30 days',
  },
  LOOKBACK_WINDOW: {
    id: 'components.mcsDateRangePicker.lookBackWindow',
    defaultMessage: 'Lookback Window',
  },
  CUSTOM: {
    id: 'components.mcsDateRangePicker.custom',
    defaultMessage: 'Custom',
  },
});

type Props = McsDateRangePickerProps & InjectedIntlProps;

class McsDateRangePicker extends React.Component<Props, McsDateRangePickerState> {
  static defaultProps: Partial<McsDateRangePickerProps> = {
    format: 'YYYY-MM-DD',
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      showRangePicker: false,
      ranges: [],
    };
  }

  componentDidMount() {
    const { excludeToday, startDate } = this.props;
    const ranges: Range[] = [
      {
        name: 'TODAY',
        from: 'now',
        to: 'now',
      },
      {
        name: 'YESTERDAY',
        from: 'now-1d',
        to: 'now-1d',
      },
    ];

    if (!startDate) {
      ranges.push(
        {
          name: 'LAST_7_DAYS',
          from: excludeToday ? 'now-8d' : 'now-7d',
          to: excludeToday ? 'now-1d' : 'now',
        },
        {
          name: 'LAST_30_DAYS',
          from: excludeToday ? 'now-31d' : 'now-30d',
          to: excludeToday ? 'now-1d' : 'now',
        },
      );
    }

    this.setState({
      ranges,
    });
  }

  disableDates = (current: moment.Moment) => {
    const { startDate } = this.props;
    return (
      current && (current.valueOf() > Date.now() || (!!startDate && current.valueOf() < startDate))
    );
  };

  getSelectedPresettedRange() {
    const { values, format, intl } = this.props;
    const { ranges } = this.state;

    const selectedRange = ranges.find(range => {
      return range.from === values.from.raw() && range.to === values.to.raw();
    });

    if (selectedRange) {
      return intl.formatMessage(messages[selectedRange.name]);
    }
    return `${convertMcsDateToMoment(values.from.raw())!.format(format)} ~ ${convertMcsDateToMoment(
      values.to.raw(),
    )!.format(format)}`;
  }

  handleDatePickerMenuChange = (dates: [moment.Moment, moment.Moment]) => {
    const { onChange } = this.props;

    this.setState({ showRangePicker: false });

    onChange({
      from: new McsMoment(dates[0].format('YYYY-MM-DD')),
      to: new McsMoment(dates[1].format('YYYY-MM-DD')),
    });
  };

  handleDropdownMenuClick = (param: MenuInfo) => {
    const { ranges } = this.state;
    const { onChange } = this.props;

    if (param.key === 'CUSTOM') {
      this.setState({
        showRangePicker: true,
      });
      return;
    }

    this.setState({ showRangePicker: false });

    const selectedRange = ranges.find(element => {
      return element.name.toLowerCase() === param.key.toString().toLowerCase();
    });

    onChange({
      from: new McsMoment(selectedRange!.from),
      to: new McsMoment(selectedRange!.to),
    });
  };

  onDatePickerOpenChange = () => {
    this.setState({
      showRangePicker: false,
    });
  };

  renderRangesDropdown() {
    const { intl, disabled } = this.props;
    const { ranges } = this.state;

    const menu = (
      <Menu onClick={this.handleDropdownMenuClick}>
        <Menu.ItemGroup title={intl.formatMessage(messages.LOOKBACK_WINDOW)}>
          {ranges.map(item => {
            return this.getSelectedPresettedRange() ===
              intl.formatMessage(messages[item.name]) ? null : (
              <Menu.Item key={item.name}>{intl.formatMessage(messages[item.name])}</Menu.Item>
            );
          })}
          <Menu.Item key='CUSTOM'>{intl.formatMessage(messages.CUSTOM)}</Menu.Item>
        </Menu.ItemGroup>
      </Menu>
    );

    return (
      <Dropdown
        className='mcs-date-range-picker'
        overlay={menu}
        trigger={['click']}
        disabled={disabled}
      >
        <Button>
          <CalendarOutlined />
          {this.getSelectedPresettedRange()}
          <DownOutlined />
        </Button>
      </Dropdown>
    );
  }

  render() {
    const { values } = this.props;
    const { showRangePicker } = this.state;
    const fromMoment = values.from.toMoment();
    const toMoment = values.to.toMoment();
    return showRangePicker ? (
      <RangePicker
        allowClear={false}
        className='mcs-date-range-picker'
        onChange={this.handleDatePickerMenuChange}
        defaultValue={[fromMoment, toMoment]}
        disabledDate={this.disableDates}
        onOpenChange={this.onDatePickerOpenChange}
        open={showRangePicker}
      />
    ) : (
      this.renderRangesDropdown()
    );
  }
}

// TODO replace any with correct type
export default injectIntl(McsDateRangePicker);
