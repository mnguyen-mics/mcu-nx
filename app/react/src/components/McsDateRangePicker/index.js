import McsDateRangePicker from './McsDateRangePicker';

export {
  McsDateRangePicker
};
