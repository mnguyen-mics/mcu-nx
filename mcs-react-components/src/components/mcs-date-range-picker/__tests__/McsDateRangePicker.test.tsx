import 'jest';
import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import McsDateRangePicker, {
  McsDateRangePickerProps,
  McsDateRangeValue,
} from '../McsDateRangePicker';
import McsMoment from '../../../utils/McsMoment';
import { IntlProvider } from 'react-intl';
jest.mock('cuid', () => () => '123');
it('should render a disabled date range picker', () => {
  const handleDatePickerMenuChange = (dates: McsDateRangeValue) => dates;
  const props: McsDateRangePickerProps = {
    values: {
      from: new McsMoment('2020-08-02'),
      to: new McsMoment('2020-08-03'),
    },
    onChange: handleDatePickerMenuChange,
    excludeToday: true,
    startDate: 1587147165831,
    disabled: true,
  };
  const component = TestRenderer.create(
    <IntlProvider locale="en">
      <McsDateRangePicker {...props} />
    </IntlProvider>,
  );
  const res = component.toJSON();
  expect(res).toMatchSnapshot();
});
it('should render a date range picker without the start date', () => {
  const handleDatePickerMenuChange = (dates: McsDateRangeValue) => dates;
  const props: McsDateRangePickerProps = {
    values: {
      from: new McsMoment('2020-08-02'),
      to: new McsMoment('2020-08-03'),
    },
    onChange: handleDatePickerMenuChange,
    excludeToday: true,
  };
  const component = TestRenderer.create(
    <IntlProvider locale="en">
      <McsDateRangePicker {...props} />
    </IntlProvider>,
  );
  const res = component.toJSON();
  expect(res).toMatchSnapshot();
});
