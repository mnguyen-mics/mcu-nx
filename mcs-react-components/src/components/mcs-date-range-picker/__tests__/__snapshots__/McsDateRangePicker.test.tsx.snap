// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a date range picker without the start date 1`] = `
<span
  id="123"
>
  <button
    className="ant-btn ant-dropdown-trigger mcs-date-range-picker"
    onClick={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    <i
      aria-label="icon: calendar"
      className="anticon anticon-calendar"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="calendar"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
        />
      </svg>
    </i>
    <span>
      2020-08-02 ~ 2020-08-03
    </span>
    <i
      aria-label="icon: down"
      className="anticon anticon-down"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="down"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
        />
      </svg>
    </i>
  </button>
</span>
`;

exports[`should render a disabled date range picker 1`] = `
<span
  id="123"
>
  <button
    className="ant-btn ant-dropdown-trigger mcs-date-range-picker"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <i
      aria-label="icon: calendar"
      className="anticon anticon-calendar"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="calendar"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
        />
      </svg>
    </i>
    <span>
      2020-08-03 ~ 2020-08-04
    </span>
    <i
      aria-label="icon: down"
      className="anticon anticon-down"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="down"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
        />
      </svg>
    </i>
  </button>
</span>
`;
